### PR TITLE
Optionally save only virtual IPs which are actually used.

### PR DESCRIPTION
**Pull request by bawr**
Reasoning: we're using vmprof in a production setting where profiles are created very frequently for specific code paths, but these don't touch ~90% of the code that is loaded, and used either before or after the hot-spot profile.

I don't expect that most users have a significant part of their profiles consist of the virtual IP section, so this is disabled by default, especially because it requires seeking through the profile data that was already written.

Still, seems like a good option to have.

Fixes an unrelated bug where some native profile records were written even if not requested, and no warning was given if native profiles *were* requested, but the stream was not readable (e.g. when the file had been opened with `wb` instead of `w+b`.

Let me know if further tweaks are needed here! ;)

### CODE REVIEW

**Comment by planrich** [comment]
Thanks for your pull request, I like the additions. I'm happy to merge it as soon as we have PyPy support!
Let me know if I can help you...

**Comment by planrich** [comment]
I'm curious for what system do you use Python + vmprof (you dont have to tell if you cant/dont want to)?

**Comment by bawr** [comment]
Is there a quick guide to testing / running vmprof on PyPy?

I haven't used it at all, so quick pointers on which version I should use and how to build / run vmprof would be most appreciated.

**Comment by planrich** [comment]
Here a quick guide:

* clone pypy repo (hg clone https://bitbucket.org/pypy/pypy), cd into it
* copy all modified C source files from vmprof-python/src/* to rpython/rlib/rvmprof/src/shared/*
* run test $ py.test pypy/module/_vmprof/test and
* run test $ py.test rpython/rlib/rvmprof/test

**Comment by bawr** [comment]
I disabled this feature on windows for now, because it heavily relies on primitive functions in symboltable, and that's only included / available on unix-like systems, it seems, so all of the common code would need to be pulled out. :(

**Comment by planrich** [comment]
The behavior is fine for Windows. Sadly Windows is a bit in a sorry state, mainly because the API Windows exposes has not been found or does not exist (for instance get a signal for a thread). Would be cool to have somebody to check again on the Windows API!

File: src/symboltable.c [hunk 1]
```diff
@@ -439,3 +439,103 @@ void dump_native_symbols(int fileno)
...
+#ifdef RPYTHON_VMPROF
+                for (i = depth/2-1; i >= 0; i--) {
+                    long kind = (long)_read_addr(fileno);
+                    void * addr = _read_addr(fileno);
+#else
+                for (i = 0; i < depth; i++) {
+                    void * addr = _read_addr(fileno);
+#endif
+                    if (!lines || (i % 2 == 1)) {
+                        LOG("found addr %p\n", addr);
```
Comment (bawr) [inline src/symboltable.c]: I know how to check for a native symbol on CPython, but do they also have that extra bit set on PyPy? If so, we can easily check teh same thing on both. If not, how do I robustly check for a native symbol on PyPy (when RPYTHON_VMPROF is defined)?

**Comment by bawr** [comment]
The PyPy tests pass. :tada: 

I went for slight indirection instead of pulling in another hashing header.
If the PyPy test runner can be taught to link with the C API, this indirection can be removed later.

**Comment by planrich** [comment]
`if (((intptr_t)addr & 0x1) == 1)` is really the check if an address is native or not. If the LSB of the pointer is one it is a native function address.

**Comment by bawr** [comment]
Right, and I'm skipping the iteration if the address happens to be native, for CPython.

For RPython I'm checking the kind - are you saying the same approach should be used for both?

**Comment by planrich** [comment]
Thanks, I have unified the two methods that walked the profile (I dont want to maintain two versions because the walking tends to get more complicated over time and different versions)

For RPython I have disabled the call to add_code_addr because it will never be called in RPython (_vmprof.c is not compiled on PyPy).

I was thinking though... I will probably enable this by default.

**Comment by bawr** [comment]
Cool! If it's going to be the default, some docs update would probably be handy, and maybe the switch could also be exposed to the write_all_code_objects method? I can make those tweaks if it's helpful.

**Comment by planrich** [comment]
Those changes are very welcome!

**Pull request merged**